plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/OpenShiftLaunchController.java

```diff
@@ -105,14 +111,36 @@ public void launch(ILaunchConfiguration configuration, String mode,
 			mode = currentMode;
 			throw e;
 		} finally {
-			checkServerState(beh, mode);
+			checkServerState(beh, currentMode, mode);
 		}
 		
 	}
 
 
-	private void checkServerState(ControllableServerBehavior beh, String mode) {
+	private void checkServerState(ControllableServerBehavior beh, String oldMode, String mode) {
 		int state = pollState();
+		
+		if (!Objects.equals(oldMode, mode)) {
+			IModule[] modules = getServer().getModules();
+			for( int i = 0; i < modules.length; i++ ) {
+				((Server)getServer()).setModulePublishState(new IModule[]{modules[i]}, IServer.PUBLISH_STATE_FULL);
+			}
+			
+			// Kick publish server job
+			Job j = new Job("Publishing server " + getServer().getName()) {
```

Comment:
no need to instantiate that job unless we're sure to use it. It should go inside the if block below